### PR TITLE
KBKDF Feedback mode + CMAC support

### PR DESCRIPTION
**Pull request by frozencemetery**
This is the final remaining KDF that Kerberos needs for modern operation: SP800-108 section 5.2 (feedback mode) using CMAC, as specified in RFC 6803.  Test vectors from RFC 6803 are included.

The first commit is a bugfix without which I can't set the cipher for CMAC.  The second could be worked around, but is inconvenient.  The third commit is the actual change.  I've grouped them since the change as written depends on the bugfixes in the first two, but could break them out into dependent PRs if that would be preferred.

krb5 test build branch (uses counter-hmac and feedback-cmac) created at https://github.com/frozencemetery/krb5/tree/800-108test

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by frozencemetery** [comment]
`-Werror=pedantic` and I are not friends.

**Comment by levitte** [comment]
Looks like a pretty straight up implementation, not at all hard to follow.

File: doc/man7/EVP_KDF-KB.pod [hunk 1]
```diff
@@ -23,21 +23,34 @@ The supported parameters are:
 
 =item B<OSSL_KDF_PARAM_PROPERTIES> ("properties") <UTF8 string>
 
-=item B<OSSL_KDF_PARAM_DIGEST> ("digest") <UTF8 string>
+=item B<OSSL_KDF_PARAM_MODE> ("mode") <UTF8 string>
```
Comment (paulidale) [inline doc/man7/EVP_KDF-KB.pod]: The format of the parameter names and types was changed recently.  These should all be of the form:
```
=item "mode" (B<OSSL_KDF_PARAM_MODE>) <UTF8 string>
```

**Comment by frozencemetery** [comment]
Parameter description seems to have missed KBKDF - added a commit to update all that and rebased on master.

File: doc/man7/EVP_KDF-KB.pod [hunk 2]
```diff
@@ -84,9 +97,37 @@ Label "label", and Context "context".
 
  EVP_KDF_CTX_free(kctx);
 
+This example derives 10 bytes using FEEDBACK-CMAC-AES256, with KI "secret",
+Label "label", and IV "sixteen bytes iv".
+
+ EVP_KDF *kdf;
```
Comment (paulidale) [inline doc/man7/EVP_KDF-KB.pod]: Most of the examples include the snippet of code to fetch the EVP_KDF and create the EVP_KDF_CTX.  Plus freeing them later.

File: providers/implementations/kdfs/kbkdf.c [hunk 3]
```diff
@@ -10,13 +10,13 @@
...
- * KDFs it defines; the name is derived from
+ * section 5.1 ("counter mode") in HMAC only and section 5.2 ("feedback mode")
+ * in CMAC only.  That document does not name the KDFs it defines; the name is
+ * derived from
  * https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Key-Derivation
  *
- * Note that sections 5.2 ("feedback mode") and 5.3 ("double-pipeline mode")
- * are not implemented, though it would be possible to do so in the future.
- * CMAC mode is also not implemented; some plumbing would be required.
+ * Note that section and 5.3 ("double-pipeline mode") is not implemented,
```
Comment (paulidale) [inline providers/implementations/kdfs/kbkdf.c]: delete the _and_.

File: test/evp_kdf_test.c [hunk 4]
```diff
@@ -297,6 +297,129 @@ static int test_kdf_x963(void)
     return ret;
 }
 
+/* KBKDF test vectors from RFC 6803 (Camellia Encryption for Kerberos 5)
+ * section 10. */
```
Comment (paulidale) [inline test/evp_kdf_test.c]: Comment formatting.  Opening `/*` and close `*/` alone on the line.

File: test/evp_kdf_test.c [hunk 5]
```diff
@@ -297,6 +297,129 @@ static int test_kdf_x963(void)
...
+         0x57, 0x7D, 0x8C, 0x37, 0x03, 0x57, 0x96, 0xAB},
+        {0x3E, 0x4F, 0xBD, 0xF3, 0x0F, 0xB8, 0x25, 0x9C,
+         0x42, 0x5C, 0xB6, 0xC9, 0x6F, 0x1F, 0x46, 0x35}
+    };
+    static unsigned char iv[16] = { 0 };
+    unsigned char result[16] = { 0 };
+
+    for (i = 0; i < 3; i++) {
+        int p = 0;
+        params[p++] = OSSL_PARAM_construct_utf8_string(
```
Comment (paulidale) [inline test/evp_kdf_test.c]: Blank line between declarations and code.

File: test/evp_kdf_test.c [hunk 6]
```diff
@@ -297,6 +297,129 @@ static int test_kdf_x963(void)
...
+         0xCD, 0x8C, 0x08, 0xE8, 0xA0, 0x24, 0x6B, 0x1D,
+         0x73, 0xB0, 0xD1, 0xDD, 0x9F, 0xC5, 0x82, 0xB0,
+        },
+    };
+    static unsigned char iv[16] = { 0 };
+    unsigned char result[32] = { 0 };
+
+    for (i = 0; i < 3; i++) {
+        int p = 0;
+        params[p++] = OSSL_PARAM_construct_utf8_string(
```
Comment (paulidale) [inline test/evp_kdf_test.c]: Blank line.

**Comment by frozencemetery** [comment]
Addressed comments and rebased on master.

File: doc/man7/EVP_KDF-KB.pod [hunk 7]
```diff
@@ -21,23 +21,36 @@ The supported parameters are:
...
+=item "info (B<OSSL_KDF_PARAM_INFO>) <octet string>
+
+=item "seed" (B<OSSL_KDF_PARAM_SEED>) <octet string>
 
 =back
 
-The parameters key, salt, and info correspond to KI, Label, and Context
-(respectively) in SP800-108.  As in that document, salt and info are optional
-and may be omitted.  Currently, only HMAC is supported for mac.
+The mode paramater determines which flavor of KBKDF to use - currently the
```
Comment (slontis) [inline doc/man7/EVP_KDF-KB.pod]: NIT: parameter

Comment (levitte) [reply doc/man7/EVP_KDF-KB.pod]: Early morning mind goes "para-matter"...  beyond normal parameters?  Strangely, it seems fitting.

I think I need more coffee...

**Comment by frozencemetery** [comment]
Minor doc update and rebase onto master.

**Comment by slontis** [comment]
Nice work.
Merged to master.
